sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -205,30 +206,145 @@ class Analyzer(
-            .map(_._2).getOrElse(u)
-        case other =>
-          // This cannot be done in ResolveSubquery because ResolveSubquery does not know the CTE.
-          other transformExpressions {
-            case e: SubqueryExpression =>
-              e.withNewPlan(substituteCTE(e.plan, cteRelations))
+    def substituteCTE(
+        plan: LogicalPlan,
+        cteRelations: Seq[(String, LogicalPlan)],
+        recursiveTableName: Option[String]): LogicalPlan = {
+      def substitute(
+          plan: LogicalPlan,
+          inSubQuery: Boolean = false): (LogicalPlan, Boolean) = {
+        val references = mutable.Set.empty[UnresolvedRecursiveReference]
+
+        def newReference(recursiveTableName: String) = {
+          val recursiveReference = UnresolvedRecursiveReference(recursiveTableName)
+          references += recursiveReference
+
+          recursiveReference
+        }
+
+        val newPlan = plan resolveOperatorsDown {
+          case u: UnresolvedRelation =>
+            val table = u.tableIdentifier.table
+
+            val recursiveReference = recursiveTableName.find(resolver(_, table)).map { name =>
+              if (inSubQuery) {
+                throw new AnalysisException(
+                  s"Recursive reference ${name} can't be used in a subquery")
+              }
+
+              newReference(name)
+            }
+
+            recursiveReference
+              .orElse(cteRelations.find(x => resolver(x._1, table)).map(_._2))
+              .getOrElse(u)
+
+          case other =>
+            // This cannot be done in ResolveSubquery because ResolveSubquery does not know the CTE.
+            other transformExpressions {
+              case e: SubqueryExpression => e.withNewPlan(substitute(e.plan, true)._1)
+            }
+        }
+
+        (newPlan, !references.isEmpty)
+      }
+
+      plan match {
+        case SubqueryAlias(name, u: Union) if recursiveTableName.isDefined =>
+          def combineUnions(union: Union): Seq[LogicalPlan] = union.children.flatMap {
+            case u: Union => combineUnions(u)
+            case o => Seq(o)
           }
+
+          val substitutedTerms = combineUnions(u).map(substitute(_))
+          val (anchorTerms, recursiveTerms) = substitutedTerms.partition(!_._2)
+
+          if (!recursiveTerms.isEmpty) {
+            if (anchorTerms.isEmpty) {
+              throw new AnalysisException("There should be at least 1 anchor term defined in a " +
+                s"recursive query $name")
+            }
+
+            val recursiveTermPlans = recursiveTerms.map(_._1)
+
+            def traversePlanAndCheck(
+                plan: LogicalPlan,
+                isRecursiveReferenceAllowed: Boolean = true): Boolean = plan match {
+              case UnresolvedRecursiveReference(name) =>
+                if (!isRecursiveReferenceAllowed) {
+                  throw new AnalysisException(s"Wrong usage of recursive reference ${name}")
```

Comment:
I think it is better if we can have reason why this usage is wrong, like Outer join is not allowed in recursive part of a recursive cte.

Reply:
I think we can have a trade-off here. I mean we can say something like:
```
Recursive reference ... cannot be used here. This can be caused by using it on the rigth side of a left-outer join or ...
```